### PR TITLE
test: make test teardown resilient to failures

### PR DESCRIPTION
**Pull request by jhockett**
#### Description of changes
Ensure child process is torn down even if test fails

#### Checklist

- [x] PR description included
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [x] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts [hunk 1]
```diff
@@ -110,16 +110,21 @@ describe('amplify pull with uibuilder', () => {
...
+      await new Promise((resolve) => setTimeout(resolve, 60000));
+      res = execa.sync(getNpxPath(), ['cypress', 'run'], { cwd: reactDir, encoding: 'utf8' });
+    } finally {
+      // kill the react server process
+      spawnSync('kill', [`${npmStartProcess.pid}`], { encoding: 'utf8' });
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+      // Seriously, kill the react server process
+      // react-scripts somehow resurrects the process automatically after the first kill.
+      spawnSync('pkill', ['-f', 'react'], { encoding: 'utf8' });
+      expect(res.exitCode).toBe(0);
```
Comment (sobolk) [inline packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:127]: should assertion go to line after `res = execa...` ??

Comment (jhockett) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:127]: Actually it should not.

This test passes
```ts
describe('jest', () => {
  it('expect in try/catch', async () => {
    try {
      expect(1).toBe(2);
    } catch {
      // blank
    }
  });
});
```


but this does not
```ts
describe('jest', () => {
  it('expect in try/catch', async () => {
    try {
      expect(1).toBe(2);
    } catch {
      // blank
    } finally {
      expect(3).toBe(4);
    }
  });
});

```

Comment (sobolk) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:127]: Yes. But... there's no `catch` block in the PR.

Comment (jhockett) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:127]: Ah, that indeed changes the behavior, my mistake

Comment (sobolk) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts]: also, consider wrapping `it` in extra `describe` block and use after test handlers for cleanup.
`try-catch-finaly` (and variations) are quite fragile due to reasons you mentioned.

**Review by sobolk — DISMISSED** [review]

**Review by sobolk — APPROVED** [review]

**Review by lazpavel — APPROVED** [review]

File: packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts [hunk 2]
```diff
@@ -110,16 +110,28 @@ describe('amplify pull with uibuilder', () => {
...
-    // kill the react server process
-    spawnSync('kill', [`${npmStartProcess.pid}`], { encoding: 'utf8' });
-    await new Promise((resolve) => setTimeout(resolve, 1000));
-    // Seriously, kill the react server process
-    // react-scripts somehow resurrects the process automatically after the first kill.
-    spawnSync('pkill', ['-f', 'react'], { encoding: 'utf8' });
-    expect(res.exitCode).toBe(0);
+    let npmStartProcess: ChildProcessWithoutNullStreams;
+    let res: execa.ExecaSyncReturnValue<string>;
+    describe('cypress tests', () => {
```
Comment (sdstolworthy) [inline packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:115]: Is this a `describe` nested inside an `it`? What purpose is this serving?

File: packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts [hunk 3]
```diff
@@ -110,16 +110,28 @@ describe('amplify pull with uibuilder', () => {
...
+      beforeAll(async () => {
```
Comment (sdstolworthy) [inline packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:116]: Same here. A lifecycle hook inside of a test block?

File: packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts [hunk 4]
```diff
@@ -110,16 +110,28 @@ describe('amplify pull with uibuilder', () => {
...
+      afterAll(async () => {
+        // kill the react server process
+        spawnSync('kill', [`${npmStartProcess.pid}`], { encoding: 'utf8' });
+        await new Promise((resolve) => setTimeout(resolve, 1000));
+        // Seriously, kill the react server process
+        // react-scripts somehow resurrects the process automatically after the first kill.
+        spawnSync('pkill', ['-f', 'react'], { encoding: 'utf8' });
+      });
+
+      it('...return exit code 0', () => {
```
Comment (sdstolworthy) [inline packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:131]:
```suggestion
      it('should return exit code 0', () => {
```

**Review by sdstolworthy — CHANGES_REQUESTED** [review]

File: packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts (hunk 2 above)
Comment (jhockett) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:115]: See convo: https://github.com/aws-amplify/amplify-cli/pull/12799#discussion_r1230048525

Comment (sobolk) [reply packages/amplify-e2e-tests/src/__tests__/uibuilder.test.ts:115]: I did not notice that we're under `it`. I think npmStartProcess can happen in `it`. The killing should go to post-it. So should `describe` wrap `it` we're now nested under instead?

**Pull request merged**